### PR TITLE
[IOTDB-163] Support create device template and create device

### PR DESCRIPTION
**Pull request by HTHou**
In the present version, it is a little trouble some to create a set timeseries that has the same measurements. On the other hand, although we use the conception "device" in the code, it is not properly abstracted.

Expected usage:

Using IoTDB in a more relational way:

CREATE COMPLEX TEMPLATE vehicle (speed DOUBLE PLAIN, direction DOUBLE PLAIN, temperature DOUBLE PLAIN, fuel DOUBLE PLAIN)

If all datatypes(or encodings) are the same, you can write the equal form:

CREATE SIMPLE TEMPLATE vehicle MEASUREMENTS (speed, direction, temperature, fuel) DATATYPE DOUBLE ENCODING PLAIN

Then you will be able to create time series in an easier way:

CREATE DEVICE (vehicle) root.sg1.vehicle1

Which equals:

CREATE TIMESERIES root.sg1.vehicle1.speed WITH DATATYPE=DOUBLE,ENCODING=PLAIN

CREATE TIMESERIES root.sg1.vehicle1.direction WITH DATATYPE=DOUBLE,ENCODING=PLAIN

CREATE TIMESERIES root.sg1.vehicle1.fuel WITH DATATYPE=DOUBLE,ENCODING=PLAIN

CREATE TIMESERIES root.sg1.vehicle1.temperature WITH DATATYPE=DOUBLE,ENCODING=PLAIN

### CODE REVIEW

**Comment by HTHou** [comment]
I can show a running result here.
Screen Shot 2019-11-07 at 10 39 25 AM

File: docs/Documentation-CHN/UserGuide/5-Operation Manual/4-SQL Reference.md [hunk 1]
```diff
@@ -147,6 +147,29 @@ Eg: IoTDB > SHOW DEVICES
 Note: This statement can be used in IoTDB Client and JDBC.
 ```
 
+* 创建设备模版语句
+我们有两种模版类型 
+简版模版用于数据类型和编码类型相同
+```
+CREATE SIMPLE TEMPLATE <TemplateName> (<Measurement> [COMMA <Measurement>]*) DATATYPE <DataTypeValue> ENCODING <EncodingValue>
```
Comment (qiaojialin) [inline docs/Documentation-CHN/UserGuide/5-Operation Manual/4-SQL Reference.md]: add a CompressionType

File: server/src/main/java/org/apache/iotdb/db/qp/executor/QueryProcessExecutor.java [hunk 2]
```diff
@@ -449,6 +457,40 @@ private boolean deleteStorageGroup(DeleteStorageGroupPlan deleteStorageGroupPlan
     }
     return true;
   }
+  
+  private boolean createDeviceTemplate(CreateDeviceTemplatePlan createDeviceTemplatePlan) 
+      throws QueryProcessException {
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/qp/executor/QueryProcessExecutor.java]:
```suggestion
{
```

File: server/src/main/java/org/apache/iotdb/db/service/TSServiceImpl.java [hunk 3]
```diff
@@ -1332,6 +1337,43 @@ public TSStatus deleteTimeseries(List<String> paths) {
     }
     return new TSStatus(executePlan(plan));
   }
+  
+  @Override
+  public TSStatus createDeviceTemplate(TSCreateDeviceTemplateReq req) {
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/service/TSServiceImpl.java]: Are these two methods used?

File: server/src/main/java/org/apache/iotdb/db/service/TSServiceImpl.java [hunk 4]
```diff
@@ -1332,6 +1337,43 @@ public TSStatus deleteTimeseries(List<String> paths) {
...
+    CreateDeviceTemplatePlan plan = new CreateDeviceTemplatePlan(req.getDeviceType(), schemaList);
+    TSStatus status = checkAuthority(plan);
+    if (status != null) {
+      return new TSStatus(status);
+    }
+    return new TSStatus(executePlan(plan));
+  }
+  
+  @Override
+  public TSStatus createDevice(TSCreateDeviceReq req) {
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/service/TSServiceImpl.java]: add an interface in session?

File: server/src/main/java/org/apache/iotdb/db/metadata/MManager.java [hunk 5]
```diff
@@ -85,6 +85,8 @@
   private long maxSeriesNumberAmongStorageGroup;
   private boolean initialized;
 
+  private Map<String, List<MeasurementSchema>> deviceTemplates = new HashMap<>();
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/metadata/MManager.java]: This should be persisted, you can refer to the persistence of mlog.txt
Recover this object when recovering

File: server/src/main/antlr3/org/apache/iotdb/db/sql/parse/TqlParser.g [hunk 6]
```diff
@@ -821,4 +860,4 @@ showTTLStatement
     |
     K_SHOW K_ALL K_TTL
     -> ^(TOK_TTL TOK_SHOW)
-    ;
\ No newline at end of file
+    ;
```
Comment (jt2594838) [inline server/src/main/antlr3/org/apache/iotdb/db/sql/parse/TqlParser.g]: The grammar should be updated to antlr4.

File: server/src/main/java/org/apache/iotdb/db/qp/executor/QueryProcessExecutor.java [hunk 7]
```diff
@@ -449,6 +457,40 @@ private boolean deleteStorageGroup(DeleteStorageGroupPlan deleteStorageGroupPlan
...
+    String deviceType = createDeviceTemplatePlan.getDeviceType();
+    List<MeasurementSchema> schemaList = createDeviceTemplatePlan.getSchemaList();
+    mManager.getDeviceTemplates().put(deviceType, schemaList);
```
Comment (jt2594838) [inline server/src/main/java/org/apache/iotdb/db/qp/executor/QueryProcessExecutor.java]: Won't this encounter any concurrency problem?

File: server/src/main/java/org/apache/iotdb/db/qp/executor/QueryProcessExecutor.java [hunk 8]
```diff
@@ -449,6 +457,40 @@ private boolean deleteStorageGroup(DeleteStorageGroupPlan deleteStorageGroupPlan
...
+        List<MeasurementSchema> schemaList = mManager.getDeviceTemplates().get(deviceType);
+        for (MeasurementSchema schema : schemaList) {
+          Path path = new Path(devicePath, schema.getMeasurementId());
+          boolean result = mManager.addPathToMTree(path, 
+              schema.getType(), schema.getEncodingType(), 
+              schema.getCompressor(), schema.getProps());
+          if (result) {
+            storageEngine.addTimeSeries(path, schema.getType(), schema.getEncodingType(), 
+                schema.getCompressor(), schema.getProps());
+          }
```
Comment (jt2594838) [inline server/src/main/java/org/apache/iotdb/db/qp/executor/QueryProcessExecutor.java]: If some of the timeseries are already created, the whole procedure will be interrupted by a MetadataException. We should discuss if it is reasonable.

File: server/src/main/java/org/apache/iotdb/db/qp/physical/sys/CreateDevicePlan.java [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    this.deviceType = deviceType;
+  }
+  
+  @Override
+  public String toString() {
+    StringBuilder stringBuilder = new StringBuilder();
+    String ret = String.format("deviceType: %s%n" + "devicePath: %s%n", 
+        deviceType, devicePath);
+    stringBuilder.append(ret);
+    return stringBuilder.toString();
```
Comment (jt2594838) [inline server/src/main/java/org/apache/iotdb/db/qp/physical/sys/CreateDevicePlan.java]: I cannot see the point of using a StringBuilder.

**Comment by qiaojialin** [comment]
https://github.com/apache/incubator-iotdb/pull/736

**Pull request closed without merging**